### ISSUE TITLE
nuget: Local feed for debugging

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#1096]
Now that we're working on sister projects in addition to neonKUBE, we really need a way to be able to work on new Neon features required by sister projects without needing to publish Neon packages to the public feed more often than we'd like.  I've been finding myself incubating entirely new libraries and code in the sister project (which will be OK for a while) and then integrating these changes back into neonKUBE and then needing to publish packages publicly so the sister project can get the changes.

We've had the `neon-nuget-local.ps` script that publishes packages to a local filesystem feed but we haven't use this because it requires the package versions to be incremented in some way so that solutions will see the updates.  The problem is that there's no nice way to increment the package versions,  Here are the problems:

1. Individual developers working in parallel will also need to increment these the version in parallel.  The current publishing mechanisms required a version file to be incremented and will update the **.csproj** files for the libraries with the new version.  We don't want to do this because **.csproj** files are tracked by git and this will result in annoying merge conflicts.

2. I was originally thinking that we'd use a special preview string like **a-####` for this, where we'd increment the build version immediately **_after_** publishing packages publicly (we currently increment this just **_before_** publishing), and then incrementing the `####` part for each local feed release.  I don't like this because it's just too easy to publicly publish packages by accident this way (has happened in the past).  The nice thing about the current scheme is that nuget won't allow to to republish packages with versions that have already been published.

   Doing this also runs in to the problems identified in **point 1** above.

3. The `neon-nuget-local.ps` script currently publishes packages to **%NF_BUILD%\nuget**.  This isn't a great place for a couple reasons: **1:** the build folder gets wiped during full builds (by script) and **2:** this implies that there will eventually be multiple local nuget feeds which will be a pain to manage.

I believe the best solution is to add a **nuget-purge** command to **neon-build** that clears specific packages from the package cache in a solution as well as from the machine/user cache at **%USERPROFILE%\.nuget\packages**.  I'll describe how we'll use this further below.

To address **point 3**, I'm going to modify the `buildenv.cmd` script to create a common local feed directory at **C:\NugetFeed** and point **%NUGET_FEED%** at it and then modify the `neon-nuget-local.ps` script to publish packages there.

I'll also modify `neon-nuget-local.ps` script to check for the presence of a `C:\NugetFeed\onpublish.ps1` script that will be called for each package version being published.  The idea here is that individual developers will add a script that uses the new **neon-build** command to remove published packages from the machine/user cache as well as the caches for the solutions the developer is working on.  I'll modify the `buildenv.cmd` to create a default script file, if one doesn't already exist.

## DEVELOPER WORKFLOW

So here's how this will work.  Let's assume that a neonKUBE developer is working on two sister solutions, one at `C:\Solution1\solution1.sln` and the other at `C:\Solution2\solution2.sln`.

### Prerequisites

1. The developer has modified the `C:\NugetFeed\onpublish.ps1` by adding the paths to the other two active solutions:
```
[CmdletBinding()]
param (
    [Parameter(Position=0, Mandatory=1)]
    [string]$packageName,
    [Parameter(Position=1, Mandatory=1)]
    [string]$packageVersion
)
neon-build nuget-purge $packageName $packageVersion "C:\Solution1" "C:\Solution2"
```

2. The developer has configured the local nuget feed to appear **before the public feed** in Visual Studio.

### Update workflow

1. The developer has made some changes to neonKUBE that need to be tested in these solutions.

2. The developer executes `neon-nuget-local.cmd` which publishes the Neon packages to the local feed directory and then calls  `C:\NugetFeed\onpublish.ps1` for each package published with this script removing any cached packages.

3. The developer will then go to each of the solutions and restore all packages (which will install the locally published packages) and do a clean rebuild.

That's all there is to it: a pretty tight inner developer loop.  Much better than the 1.5-2 hours it takes to do a public release these days and we'll also avoid confusing the community with too many releases.

**Comment by jefflill** [comment]
NuGet feeds are not working as I expected.  Although Visual Studio looks like it allows you to order how the nuget feeds are searched, that doesn't actually work.  Apparently, it queries all feeds in parallel and retrieves packages from the feed that returns first (although you'd think that local feeds would always return first and they don't seem to).

I was hoping that by removing the locally published packages from the machine caches and having the local feed first in the list that I'd be able to republish packages locally without updating the version and that other solutions would then pick up the new package **without having to upgrade packages**.  This doesn't to work because VS seems to favor **nuget.org** over local feeds and we always get the original package,  **nuget** fundamentally assumes that packages are invariant after being published, so the developers say that it shouldn't matter where a package comes from since it should always be the same.

I don't think there's really a way to accomplish this cleanly with packages.  The only remaining options are:

1. Disable the **nuget.org** package feed while working in dependent solutions.

2. Increasing the version number for locally published packages to something like **1000.0.0-####**, where #### is incremented for every time packages are published locally.  Developers will need to update packages in their solutions to pick these up.

3. Have the dependent solutions reference the (probably debug) builds of the Neon library assemblies directly while developing and debugging and then restoring the package referenced afterwards.

4. Include neonKUBE library projects in the sister solutions.  We'll need to be careful to replace these direct project references with package reference before merging into master and especially before publish the neonLIBRARY packages.  I'm not entirely convinced that this will work.

**Option 2** is probably the best choice.  By auto incrementing the preview tag version, we won't be subverting the nuget package impotence requirement and the only requirement for developers will be to upgrade their solution package references to use the locally published packages and the be sure to downgrade the packages back to the release versions before merging to master or release branches.

One consequence of managing this tag version locally is that different developers will end up with different tag versions.  This means that developers won't be able to work with locally published packages on the same branch.  This sucks, but I can't think of anything better outside of maintaining our own NuGet server and managing the version number there too.

I'm going to go with **Option 2** without deploying our own NuGet server at first.  I'm not entirely sure how often we'll actually be doing this anyway.  If this ends up being important, we'll just implement our own package server which will be as clean is this possible can be.

**Comment by jefflill** [comment]
DONE
